service/history/replication/executable_activity_state_task.go

```diff
@@ -209,3 +240,54 @@ func (e *ExecutableActivityStateTask) MarkPoisonPill() error {
 
 	return writeTaskToDLQ(ctx, e.DLQWriter, shardContext, e.SourceClusterName(), replicationTaskInfo)
 }
+
+func (e *ExecutableActivityStateTask) BatchWith(incomingTask BatchableTask) (TrackableExecutableTask, bool) {
+	if !e.batchable || !incomingTask.CanBatch() {
+		return nil, false
+	}
+	e.batchLock.Lock()
+	defer e.batchLock.Unlock()
+
+	if !e.batchable || !incomingTask.CanBatch() {
+		return nil, false
+	}
+	
+	incomingActivityTask, err := e.validateIncomingBatchTask(incomingTask)
+	if err != nil {
+		return nil, false
+	}
+	return &ExecutableActivityStateTask{
+		ProcessToolBox: e.ProcessToolBox,
+		WorkflowKey:    e.WorkflowKey,
+		ExecutableTask: e.ExecutableTask,
+		batchable:      true,
+		activityInfos:  append(e.activityInfos, incomingActivityTask.activityInfos...),
+	}, true
+}
+
+func (e *ExecutableActivityStateTask) validateIncomingBatchTask(incomingTask BatchableTask) (*ExecutableActivityStateTask, error) {
+	if incomingTask == nil {
+		return nil, serviceerror.NewInvalidArgument("Batch task is nil")
+	}
+	incomingActivityTask, isActivityTask := incomingTask.(*ExecutableActivityStateTask)
+	if !isActivityTask {
+		return nil, serviceerror.NewInvalidArgument("Unsupported Batch type")
+	}
+	if e.WorkflowKey != incomingActivityTask.WorkflowKey {
+		return nil, serviceerror.NewInvalidArgument("WorkflowKey mismatch")
+	}
+
+	return incomingActivityTask, nil
+}
+
+func (e *ExecutableActivityStateTask) CanBatch() bool {
+	e.batchLock.Lock()
+	defer e.batchLock.Unlock()
+	return e.batchable
+}
+
+func (e *ExecutableActivityStateTask) MarkUnbatchable() {
+	e.batchLock.Lock()
+	defer e.batchLock.Unlock()
+	e.batchable = false
```

Comment:
Batchable should only be accessed within the lock or outside of the lock as a sync/atomic.Bool or something. We're mixing access patterns here